programs/stake/src/stake_state.rs

```diff
@@ -856,6 +856,113 @@ pub fn merge(
+            invoke_context,
+            "expected len {}, got len {}",
+            StakeState::size_of(),
+            uninitialized_stake_account.get_data().len()
+        );
+        return Err(InstructionError::InvalidAccountData);
+    }
+    if !matches!(
+        uninitialized_stake_account.get_state()?,
+        StakeState::Uninitialized
+    ) {
+        return Err(InstructionError::InvalidAccountData);
+    }
+
+    // validate the provided vote account
+    let vote_account = instruction_context
+        .try_borrow_instruction_account(transaction_context, vote_account_index)?;
+    if *vote_account.get_owner() != solana_vote_program::id() {
+        return Err(InstructionError::IncorrectProgramId);
+    }
+    let vote_pubkey = *vote_account.get_key();
+    let vote_state = vote_account.get_state::<VoteStateVersions>();
+
+    // ensure `stake_account` is active, check for presence of the stake authority, and
+    // extract its Meta
+    let meta = if let StakeState::Stake(meta, stake) = stake_account.get_state()? {
+        meta.authorized.check(signers, StakeAuthorize::Staker)?;
+
+        let status = stake
+            .delegation
+            .stake_activating_and_deactivating(clock.epoch, Some(stake_history));
+        if status.effective == 0 || status.activating != 0 || status.deactivating != 0 {
+            ic_msg!(invoke_context, "stake is not active");
+            return Err(InstructionError::InvalidAccountData);
+        }
+
+        // Deny redelegating to the same vote account. This is nonsensical and could be used to
+        // grief the global stake warm-up/cool-down rate
+        if stake.delegation.voter_pubkey == vote_pubkey {
+            ic_msg!(
+                invoke_context,
+                "redelegating to the same vote account not permitted"
+            );
+            return Err(InstructionError::InvalidArgument);
+        }
+
+        meta
+    } else {
+        return Err(InstructionError::InvalidAccountData);
+    };
+
+    // deactivate `stake_account`
+    deactivate(stake_account, clock, signers)?;
+
+    // transfer all but the `rent_exempt_reserve` balance to the uninitialized stake account
+    let redelegation_amount = stake_account
+        .get_lamports()
+        .checked_sub(meta.rent_exempt_reserve)
+        .ok_or(InstructionError::InsufficientFunds)?;
+    stake_account.checked_sub_lamports(redelegation_amount)?;
+    uninitialized_stake_account.checked_add_lamports(redelegation_amount)?;
+
+    // initialize and delegate `uninitialized_stake_account`
+    let ValidatedDelegatedInfo { stake_amount } = validate_delegated_amount(
+        &uninitialized_stake_account,
+        &meta,
+        &invoke_context.feature_set,
+    )?;
+    let stake = new_stake(
+        stake_amount,
+        &vote_pubkey,
+        &vote_state?.convert_to_current(),
+        clock.epoch,
+        config,
+    );
+    uninitialized_stake_account.set_state(&StakeState::Stake(meta, stake))?;
```

Comment:
I'm not sure if any still exist, but legacy stakes with a huge rent-exempt reserve because of an overallocation would be confusing in this case.  For example, if I redelegate from a stake with 2000 bytes to a stake with 200 bytes, it would have the old stake's larger rent-exempt reserve

Reply:
Oh interesting. I guess that's ok. The size of `uninitialized_stake_account` is required to be `StakeState::size_of()`, so worse case slightly more lamports are deactivated if a megalodon legacy stake account is provided as input. It doesn't seem like we'd need to do anything about this.

Reply:
It's a pretty easy fix to update the rent exempt amount in the meta to whatever the rent exempt amount is in `uninitialized_stake_account`, no?

Reply:
Yep, just slightly more code. Here's what that looks like: 007371b6b2c2c7017732c0bcb43d8c520c5c060a